### ISSUE TITLE
Understanding the implementation of the Integrated Gradient

### ISSUE THREAD

**Issue by chris-tkinter** [pytorch/captum#713]
## ❓ Questions and Help

Hi, I would like to thank you for making this powerful library!

I have a question regarding the implementation of the Integrated Gradient (IG). The original paper mentions the approximation of the path integral by 
![image](https://user-images.githubusercontent.com/86935669/125137321-4a03b900-e0c1-11eb-9d71-1404d4e05b22.png)

In the code I am wondering about this [line](https://github.com/pytorch/captum/blob/master/captum/_utils/gradient.py#L119) 
```python
grads = torch.autograd.grad(torch.unbind(outputs), inputs)
```
Let's ignore the details of multiple output & input and let's say we have an input of shape `(2, 3)` where 2 is the batch size. The captum would generate scaled input for the step_size (the default is 50) so the new input (and the `inputs` in the above code) would be `(100, 3)`, where `inputs[n]` would mean the ith sample in the batch with jth step for some i and j. And let's say the `outputs` would be of shape `(100, 2)` where `outputs[n]` would mean the output of the `inputs[n]`. Later we do a slicing on the `outputs` and make it `(100, )` shape.

I hope that what I wrote above is right.

Let's say the unbinded outputs is `(l1, ..., ln)`. Then for this line, `grad((l1, l2, ..., ln), inputs)` would return grad1 + grad2 + .. + gradn where gradi = gradient of li w.r.t. inputs. The `grads` has shape `(100, 3)` so I think what this code does is to find the gradient of `x' + k/m (x-x')` w.r.t. `x_i`. But I am wondering why we sum over all `li` here. If `inputs[n]` is the ith sample in the batch with **jth** step for some i and j, why would it sum some lz which is the output of some **zth** step?

Thank you!

**Comment by NarineK** [comment]
@chris-tkinter, the equation represents IG computation for one example. We adjust it for multiple example in one batch and as you mentioned, if inputs is (batch_size x #features), then after interpolation from x_prime to x with the #steps, we get (batch_size * num_steps  x #features).  `num_steps` examples for the first input example followed by `num_steps` examples for the second input example. 

Gradients are computed for the `x' + k/m (x-x')` w.r.t. `x_i`  as you mentioned. `x' + k/m (x-x')` is a single valued output for each `x_i` 
see comment here:
https://github.com/pytorch/captum/blob/master/captum/attr/_core/integrated_gradients.py#L366

hence there is no summation s.a.  grad1 + grad2 + .. + gradn. As you can see the gradients have the same shape as the inputs. 

The actual and only summation that you see in the equation is happening here: 
https://github.com/pytorch/captum/blob/master/captum/attr/_core/integrated_gradients.py#L366

I'm not sure that I understood your question correctly but I hope that this helps.

**Comment by chris-tkinter** [comment]
@NarineK Thank you for your reply! 
> hence there is no summation s.a. grad1 + grad2 + .. + gradn. As you can see the gradients have the same shape as the inputs.

I think I understand that part, so the line here is for the sum from k = 1 to m, is it correct?

When I mentioned the grad1 + grad2 + .. I meant what the `torch.autograd.grad` returns, from the [doc](https://pytorch.org/docs/stable/generated/torch.autograd.grad.html#torch.autograd.grad)

> Computes and returns the sum of gradients of outputs with respect to the inputs.

As an example

```python
import torch
a = torch.tensor(2., requires_grad=True)
b = torch.tensor(5., requires_grad=True)
z = 3*a + 2*b
z2 = 6*a + 9*b
grad((z, z2), a)
# output (9, )
```
here the output is 9 since it is the grad of z1 wrt a + grad of z2 wrt a. so 3 + 6. Going back to the line  
```python
grads = torch.autograd.grad(torch.unbind(outputs), inputs)
```
If the unbinded outputs is `(l1, ..., ln)`, it would, I think, sum each gradi together, where gradi is the gradient of li w.r.t. inputs.

Therefore I was a little unsure about what this line does here, since it doesn't show up in the paper's formula. I hope that this question is a clearer now. Thanks!

**Comment by NarineK** [comment]
@chris-tkinter, thank you for providing an example. What you are doing is something else. You have two different functions, and you are computing the gradients of outputs generated by two different functions w.r.t to an input. In this case you will get summation.
We are working with one function. There is one neural network. So , what we do can be described as:

```
import torch
a = torch.tensor([[2.],[2.]], requires_grad=True) # 2 examples, 1 feature - 2x1
b = torch.tensor([[5.],[5.]], requires_grad=True) # 2 examples, 1 feature - 2x1
z = 3*a + 2*b # 2 x 1, one output per example, not two functions per single example
torch.autograd.grad(torch.unbind(z), a)

```
output: 
(tensor([[3.],
         [3.]]),)

**Comment by chris-tkinter** [comment]
Oh I see! Thank you for your detailed explanation! That makes perfect sense now! Closing this issue now.
